### PR TITLE
reduce number of health checks 

### PR DESCRIPTION
**Pull request by katdrobnjakovic**

### CODE REVIEW

**Comment by katdrobnjakovic** [comment]
I ran my changes in nginx-routing-modules and the tests for monitor_upstreams and traffic_status pass:

```
katdrobnjakovic@Kats-MacBook-Pro-2 nginx-routing-modules (health-checks) $ dev test test/integration/monitor_upstreams_test.rb 
Started with run options --seed 51663

  6/6: [=====================================================================================================================================] 100% Time: 00:00:02, Time: 00:00:02

Finished in 2.68917s
6 tests, 108 assertions, 0 failures, 0 errors, 0 skips
katdrobnjakovic@Kats-MacBook-Pro-2 nginx-routing-modules (health-checks) $ dev test test/integration/traffic_status_test.rb 
Started with run options --seed 7113

  3/3: [=====================================================================================================================================] 100% Time: 00:00:00, Time: 00:00:00

Finished in 0.20149s
3 tests, 42 assertions, 0 failures, 0 errors, 0 skips
```

File: lib/resty/upstream/healthcheck.lua [hunk 1]
```diff
@@ -536,6 +522,92 @@ local function preprocess_peers(peers)
...
+end
+
+local function get_checked_peer(peer_name)
+    return checked_peers[peer_name]
+end
+
+local function get_all_upstream_peers(u)
+    local primary_peers, err = upstream.get_primary_peers(u)
+    if not primary_peers then
+        ngx.say("failed to get peers: ", err)
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: This should probably write to the error log.

Comment (csfrancis) [reply lib/resty/upstream/healthcheck.lua]: Maybe also be descriptive about which peer list had an error (primary vs. backup).

File: lib/resty/upstream/healthcheck.lua [hunk 2]
```diff
@@ -536,6 +522,92 @@ local function preprocess_peers(peers)
...
+    local ppeers_to_check = {}
+    local bpeers_to_check = {}
+
+    for k, ppeer in pairs(primary_peers) do
+        if get_checked_peer(ppeer.name) == nil then
+            ppeers_to_check[k] = ppeer
+        end
+    end
+
+    for k, bpeer in pairs(backup_peers) do
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: This `for` block is pretty much identical to the previous - might be worth refactoring into a common function.

File: lib/resty/upstream/healthcheck.lua [hunk 3]
```diff
@@ -536,6 +522,92 @@ local function preprocess_peers(peers)
...
+    local ppeers_checked = {}
+    local bpeers_checked = {}
+
+    for k, ppeer in pairs(primary_peers) do
+        if get_checked_peer(ppeer.name) ~= nil then
+            ppeers_checked[k] = ppeer
+        end
+    end
+
+    for k, bpeer in pairs(backup_peers) do
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: Same comment as in `get_upstream_unique_peers`. These `for` blocks could be in a common function.

File: lib/resty/upstream/healthcheck.lua [hunk 4]
```diff
@@ -289,85 +294,66 @@ local function check_peer(ctx, id, peer, is_backup)
...
-                if rest == 0 then
-                    break
+            while n > 0 do
+                for k, peer in pairs(peers) do
+                    if debug_mode then
+                        debug("spawn a thread checking ",
+                            is_backup and "backup" or "primary", " peer ",
+                            peer.id)
+                    end
+                    threads[k] = spawn(check_peer, ctx, peer.id, peer, is_backup)
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: You need to be calling `check_peer_range` here. Otherwise you're breaking concurrency limit specified by `concur`.

Comment (csfrancis) [reply lib/resty/upstream/healthcheck.lua]: Also, L337 precomputes the number of threads based on the group size, which is now being exceeded. This means that on L361 you won't actually end up waiting for all the health check threads to complete.

File: lib/resty/upstream/healthcheck.lua [hunk 5]
```diff
@@ -289,85 +294,66 @@ local function check_peer(ctx, id, peer, is_backup)
...
-local function check_peer_range(ctx, from, to, peers, is_backup)
-    for i = from, to do
-        check_peer(ctx, i - 1, peers[i], is_backup)
-    end
-end
-
 local function check_peers(ctx, peers, is_backup)
-    local n = #peers
+    local n = 0
+    for _ in pairs(peers) do n = n + 1 end
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: Why can't you just use `#peers`?

File: lib/resty/upstream/healthcheck.lua [hunk 6]
```diff
@@ -650,6 +729,10 @@ function _M.spawn_checker(opts)
 
     update_upstream_checker_status(u, true)
 
+    if dedup == true then
+        copy_status(ctx)
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: This is not doing what you think - it will only call `copy_status` on the first iteration, and not for subsequent health checks.

File: lib/resty/upstream/healthcheck.lua [hunk 7]
```diff
@@ -150,6 +154,7 @@ end
 
 local function peer_ok(ctx, is_backup, id, peer)
     debug("peer ", peer.name, " was checked to be ok")
+    peer_status[peer.name] = "ok"
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: I don't think this will work as you expect. The peer is not always marked as `up` by this function. If the peer is marked as up by `upgrade_peers_version` (which happens for on every worker that did not actually run the health check request), the `peer_status` table won't be updated.

**Review by csfrancis — DISMISSED** [review]

File: lib/resty/upstream/healthcheck.lua (hunk 5 above)
Comment (katdrobnjakovic) [inline lib/resty/upstream/healthcheck.lua]: Because peers is a hash and not an array, after the first upstream, for the next upstream #peers returns 0.

File: lib/resty/upstream/healthcheck.lua (hunk 6 above)
Comment (katdrobnjakovic) [reply lib/resty/upstream/healthcheck.lua]: Do you mean first iteration as in first upstream that is checked?

File: lib/resty/upstream/healthcheck.lua (hunk 3 above)
Comment (katdrobnjakovic) [reply lib/resty/upstream/healthcheck.lua]: done

File: lib/resty/upstream/healthcheck.lua (hunk 2 above)
Comment (katdrobnjakovic) [reply lib/resty/upstream/healthcheck.lua]: done

File: lib/resty/upstream/healthcheck.lua (hunk 1 above)
Comment (katdrobnjakovic) [reply lib/resty/upstream/healthcheck.lua]: done

File: lib/resty/upstream/healthcheck.lua (hunk 4 above)
Comment (katdrobnjakovic) [reply lib/resty/upstream/healthcheck.lua]: ive changed it back to using check_peer_range and tophatted and it looks good. the conditional that was failing in the tophat before was when #peers <= concurrency

File: lib/resty/upstream/healthcheck.lua (hunk 6 above)
Comment (csfrancis) [reply lib/resty/upstream/healthcheck.lua]: Yeah. When `check` is called again, `copy_status` won't be called.

File: lib/resty/upstream/healthcheck.lua [hunk 8]
```diff
@@ -296,39 +298,43 @@ local function check_peer_range(ctx, from, to, peers, is_backup)
...
+        for _, peer in pairs(peers) do
+            check_peer(ctx, peer.id, peer, is_backup)
         end
     else
         local threads
         local nthr
 
         if n <= concur then
-            nthr = n - 1
+            nthr = n
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: Why did you change this?

File: lib/resty/upstream/healthcheck.lua [hunk 9]
```diff
@@ -296,39 +298,43 @@ local function check_peer_range(ctx, from, to, peers, is_backup)
...
             threads = new_tab(nthr, 0)
-            for i = 1, nthr do
+
+            for k, peer in pairs(peers) do
+                if next(peers, _) == nil then
```
Comment (csfrancis) [inline lib/resty/upstream/healthcheck.lua]: This looks really strange to me - what are you trying to do here?

Comment (katdrobnjakovic) [inline lib/resty/upstream/healthcheck.lua]: checking the last peer with the current thread. 
because `peers` cant be indexed `peers[n]` returns nil

File: lib/resty/upstream/healthcheck.lua [hunk 10]
```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+    local checked_peers = {}
+    for _, peer in ipairs(peers) do
+        if checked_peers_upstream[peer.name] then
+            table.insert(checked_peers, peer)
+        end
+    end
+    return checked_peers
+end
+
+local function get_all_upstream_peers(u)
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: Function names should be consistent. Currently, we have: `get_all_upstream_peers`, `get_unique_upstream_peers`, `already_checked_peers` returning a pair of arrays (primary & backup), and `get_peers_to_check` and `get_checked_peers` returning a single filtered array. `already_checked_peers` returns the result of a pair of calls to `get_checked_peers` and `get_unique_upstream_peers` returns the result of a pair of calls to `get_peers_to_check`.
The following renames would be clearer:
- `get_all_upstream_peers -> get_all_peers`
- `get_unique_upstream_peers -> get_all_unchecked_peers`
- `get_peers_to_check  -> get_unchecked_peers`
- `already_checked_peers -> get_all_checked_peers`

File: lib/resty/upstream/healthcheck.lua [hunk 11]
```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+    if not backup_peers then
+        errlog("failed to get backup peers: ", err)
+        return
+    end
+
+    return primary_peers, backup_peers
+end
+
+local function get_unique_upstream_peers(u)
+    primary_peers, backup_peers = get_all_upstream_peers(u)
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: These should be `local`.

File: lib/resty/upstream/healthcheck.lua [hunk 12]
```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+
+    local ppeers_to_check = get_peers_to_check(primary_peers)
+    local bpeers_to_check = get_peers_to_check(backup_peers)
+
+    return ppeers_to_check, bpeers_to_check
+end
+
+local function already_checked_peers(u)
+    primary_peers, backup_peers = get_all_upstream_peers(u)
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: These should be `local`.

File: lib/resty/upstream/healthcheck.lua [hunk 13]
```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+
+        local noks, err = dict:get(gen_peer_key("nok:", checked_peer_upstream, is_backup, peer.id))
+        if noks then
+            dict:set(gen_peer_key("nok:", u, is_backup, peer.id), 1)
+        end
+    end
+end
+
+local function copy_status(ctx)
+    ppeers_checked, bpeers_checked = already_checked_peers(ctx.upstream)
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: These should be `local`

File: lib/resty/upstream/healthcheck.lua [hunk 14]
```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+    for i = 1, #peers do
+        local peer = peers[i]
+        local already_set = dict:get(gen_peer_key("ok:", u, is_backup, peer.id))
+        if already_set then
+            return
+        end
+
+        local checked_peer_upstream = checked_peers_upstream[peer.name]
+
+        local oks, err = dict:get(gen_peer_key("ok:", checked_peer_upstream, is_backup, peer.id))
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: What happens if `err` is set here? If it matters, do something with it. If it doesn't, then `local oks, _ = ...`

File: lib/resty/upstream/healthcheck.lua [hunk 15]
```diff
@@ -613,14 +704,21 @@ function _M.spawn_checker(opts)
...
-    if not ppeers then
-        return nil, "failed to get primary peers: " .. err
-    end
+    local dedup = opts.dedup
+
+    local ppeers, bpeers
+    if dedup then
+        ppeers, bpeers = get_unique_upstream_peers(u)
+
+        for i = 1, #ppeers do
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: `for _, peer in ipairs(ppeers) do`

File: lib/resty/upstream/healthcheck.lua [hunk 16]
```diff
@@ -613,14 +704,21 @@ function _M.spawn_checker(opts)
...
+            checked_peers_upstream[ppeers[i].name] = u
+        end
 
-    local bpeers, err = get_backup_peers(u)
-    if not bpeers then
-        return nil, "failed to get backup peers: " .. err
+        for i = 1, #bpeers do
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: `for _, peer in ipairs(bpeers) do`

**Review by fbogsany — CHANGES_REQUESTED** [review]

File: lib/resty/upstream/healthcheck.lua [hunk 17]
```diff
@@ -613,14 +704,21 @@ function _M.spawn_checker(opts)
...
 
-    local ppeers, err = get_primary_peers(u)
-    if not ppeers then
-        return nil, "failed to get primary peers: " .. err
-    end
+    local dedup = opts.dedup
+
+    local ppeers, bpeers
+    if dedup then
+        ppeers, bpeers = get_all_unchecked_peers(u)
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: This is the only place you're calling `get_all_unchecked_peers`, and that function is the only caller of `get_unchecked_peers`. Here you're setting `checked_peers_upstream[peer.name] = u` for all unchecked peers. It seems more succinct to:
```lua
local ppeers, bpeers = get_all_peers(u)
if dedup then
  ppeers = select_and_check_unchecked_peers(ppeers, u)
  bpeers = select_and_check_unchecked_peers(bpeers, u)
end
```
here, with:
```lua
local function select_and_check_unchecked_peers(peers, u)
  local filtered_peers = {}
  for _, peer in ipairs(peers) do
    if not checked_peers_upstream[peer.name] then
      checked_peers_upstream[peer.name] = u
      table.insert(filtered_peers, peer)
    end
  end
  return filtered_peers
end
```
If there's overlap between backup and primary peer names (e.g. some primary peers appear as backup peers for other upstreams), then the result is slightly different from your code. If the distinction is important, it's easy to modify `select_and_check_unchecked_peers` to take `(ppeers, bpeers, u)` and loop over both arrays, filtering first, then setting.

File: lib/resty/upstream/healthcheck.lua [hunk 18]
```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+        end
+    end
+end
+
+local function copy_status(ctx)
+    local ppeers_checked, bpeers_checked = get_all_checked_peers(ctx.upstream)
+
+    set_status(ppeers_checked, false, ctx)
+    set_status(bpeers_checked, true, ctx)
+end
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: This function is the only place you call `get_all_checked_peers`, which is the only caller of `get_checked_peers`. Again, this could be simplified by inlining the ... uh ... check into `set_status`, which should probably be called `copy_status_if_checked`:
```lua
local function copy_status(ctx)
  local ppeers, bpeers = get_all_peers(ctx.upstream)
  copy_status_if_checked(ppeers, false, ctx)
  copy_status_if_checked(bpeers, true, ctx)
end
```

File: lib/resty/upstream/healthcheck.lua [hunk 19]
```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+local function set_status(peers, is_backup, ctx)
+    local dict = ctx.dict
+    local u = ctx.upstream
+
+    for i = 1, #peers do
+        local peer = peers[i]
+        local already_set = dict:get(gen_peer_key("ok:", u, is_backup, peer.id))
+        if already_set then
+            return
+        end
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: This then becomes:
```lua
local function copy_status_if_checked(peers, is_backup, ctx)
  local dict = ctx.dict
  local u = ctx.upstream

  for _, peer in ipairs(peers) do
    local id = peer.id
    local checked_peer_upstream = checked_peers_upstream[peer.name]
    -- Is it sufficient to only check "ok:..."? It's possible we only set "nok:...".
    if checked_peer_upstream and not dict:get(gen_peer_key("ok:", u, is_backup, id)) then
      copy_peer_status(dict, "ok:", checked_peer_upstream, is_backup, id, u)
      copy_peer_status(dict, "nok:", checked_peer_upstream, is_backup, id, u)
    end
  end
end
```
where:
```lua
local function copy_peer_status(dict, prefix, checked_upstream, is_backup, id, upstream)
  local val, _ = dict:get(gen_peer_key(prefix, checked_upstream, is_backup, id))
  if val then
    dict:set(gen_peer_key(prefix, upstream, is_backup, id), 1)
  end
end
```

File: lib/resty/upstream/healthcheck.lua [hunk 20]
```diff
@@ -92,6 +93,92 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+
+local function set_status(peers, is_backup, ctx)
+    local dict = ctx.dict
+    local u = ctx.upstream
+
+    for i = 1, #peers do
+        local peer = peers[i]
+        local already_set = dict:get(gen_peer_key("ok:", u, is_backup, peer.id))
+        if already_set then
+            return
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: The early return here is incorrect (afaict) - it skips processing the remaining peers as soon as it finds one that is set. I think you want a `continue` here, but Lua doesn't provide that.

**Review by fbogsany — CHANGES_REQUESTED** [review]

File: lib/resty/upstream/healthcheck.lua (hunk 20 above)
Comment (fbogsany) [reply lib/resty/upstream/healthcheck.lua]: Read the following comments in reverse order 😄 . Github was unhelpful here.

File: lib/resty/upstream/healthcheck.lua [hunk 21]
```diff
@@ -92,6 +93,71 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+
+        local noks, _ = dict:get(gen_peer_key("nok:", checked_peer_upstream, is_backup, id))
+        if noks then
+            dict:set(gen_peer_key("nok:", u, is_backup, id), 1)
+        end
+    end
+end
+
+local function copy_status(ctx)
+    local ppeers_checked, bpeers_checked = get_all_peers(ctx.upstream)
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: These aren't necessarily checked yet, so it's more correct to call them `ppeers, bpeers`.

File: lib/resty/upstream/healthcheck.lua [hunk 22]
```diff
@@ -92,6 +93,71 @@ local function set_peer_down_globally(ctx, is_backup, id, value)
...
+  return filtered_peers
+end
+
+local function copy_status_if_checked(peers, is_backup, ctx)
+    local checked_peers = {}
+    for _, peer in ipairs(peers) do
+        if checked_peers_upstream[peer.name] then
+            table.insert(checked_peers, peer)
+        end
+    end
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: This loop doesn't do anything. The `checked_peers` table isn't used.

**Review by fbogsany — CHANGES_REQUESTED** [review]

**Comment by katdrobnjakovic** [comment]
Ive updated this with a few things: 
- the peers are now marked as checked once the check is complete. Previously they were marked as checked before the check was completed which could cause a race condition if they are marked as checked and then that worker dies before the actual check is completed. 
- to copy the status I now call `peer_ok` or `peer_fail`. Previously I was only copying the `ok` and `nok` part of the status but this didnt include the increment for `check_ok` or `check_fail` so they would always show up in the status table as 1.

**Review by es — APPROVED** [review]
I don't think we should make de-duping optional, but that can be safely removed in a follow-up PR without much hassle. 

🚀

File: lib/resty/upstream/healthcheck.lua [hunk 23]
```diff
@@ -209,9 +241,42 @@ local function peer_error(ctx, is_backup, id, peer, ...)
...
+    local dict = ctx.dict
+    local u = ctx.upstream
+
+    local id = peer.id
+    local set_in_current_upstream = dict:get(gen_peer_key("ok:", u, is_backup, id))
+
+    local checked_peer_upstream = dict:get('checked_peers_upstream:' .. peer.name)
+    local peer_set_in_different_upstream = dict:get(gen_peer_key("ok:", checked_peer_upstream, is_backup, id))
+
+
```
Comment (fbogsany) [inline lib/resty/upstream/healthcheck.lua]: Extra blank line.

**Review by fbogsany — APPROVED** [review]
Nitpicky whitespace, but LGTM

**Review by fbogsany — APPROVED** [review]

File: t/sanity.t [hunk 24]
```diff
@@ -1966,3 +1968,479 @@ init_worker_by_lua_block {
...
+--- config
+    location = /t {
+        access_log off;
+        content_by_lua_block {
+            ngx.sleep(0.52)
+            local hc = require "resty.upstream.healthcheck"
+            local status_table = hc.status_table("healthcheck")
+
+            local p12355 = status_table["foo.com"].primary_peers[1]
+            if p12355.checks_ok ~= 7 or p12355.unhealthy ~= false or p12355.checks_fail ~= 0 then
```
Comment (ElvinEfendi) [inline t/sanity.t:2423]: Should this not be 6 instead of 7? Because the interval is `0.1` second and the block sleeps `0.52` seconds(`6 = (0.52/0.1).ceil`).

File: t/sanity.t [hunk 25]
```diff
@@ -1966,3 +1968,479 @@ init_worker_by_lua_block {
...
+healthcheck: peer 127.0.0.1:12355 was checked to be ok
+healthcheck: peer 127.0.0.1:12357 was checked to be ok
+healthcheck: peer 127.0.0.1:12358 was checked to be ok
+healthcheck: peer 127.0.0.1:12356 was checked to be ok
+healthcheck: peer 127.0.0.1:12359 was checked to be ok
+
+--- timeout: 6
+
+
+=== TEST 27: status_table should return peers with correct properties
```
Comment (ElvinEfendi) [inline t/sanity.t:2361]: This title is duplicate of the title for `TEST 23`. Maybe append " when dedup is true"?

File: lib/resty/upstream/healthcheck.lua [hunk 26]
```diff
@@ -209,9 +241,41 @@ local function peer_error(ctx, is_backup, id, peer, ...)
...
+    end
+
+    local noks, _ = dict:get(gen_peer_key("nok:", checked_peer_upstream, is_backup, id))
+    if noks then
+        peer_fail(ctx, is_backup, id, peer)
+    end
+end
+
+local function propogate_peer_status(ctx)
+    local ppeers, bpeers = get_all_peers(ctx.upstream)
```
Comment (ElvinEfendi) [inline lib/resty/upstream/healthcheck.lua:265]: During initialization we fetch peers into `ctx` table. I think we should use them here instead of refetching peers:
```
local ppeers, bpeers = ctx.primary_peers, ctx.backup_peers
```

**Review by ElvinEfendi — CHANGES_REQUESTED** [review]
I could not find how 
```
'checked_peers_upstream:' .. peer.name
```
gets invalidated. If you don't invalidate this no future health check will happen even after the given interval has passed. I think
```
dict:set('checked_peers_upstream:' .. peer.name, u)
```
should be done with expiration, i.e something like this:
```
dict:set('checked_peers_upstream:' .. peer.name, u, ctx.interval)
```

File: lib/resty/upstream/healthcheck.lua (hunk 26 above)
Comment (es) [reply lib/resty/upstream/healthcheck.lua:265]: Ideally, we should be minimizing the number of reads we do directly on the `ctx` object as it's very expensive.

Comment (ElvinEfendi) [reply lib/resty/upstream/healthcheck.lua:265]: @es `ctx` is not a global variable like `ngx.ctx` here. It's defined locally at https://github.com/Shopify/lua-resty-upstream-healthcheck/pull/6/files#diff-49d6ee97295edb7b532cf5930cfb8b5eR698 and passed by reference to this function. Therefore I don't think there's any performance penalty in reading it's properties.

Comment (es) [reply lib/resty/upstream/healthcheck.lua:265]: You're absolutely right :man_facepalming:

**Comment by marc-barry** [comment]
We are approaching 1 million RPM in health checks. It would be nice to get this shipped so we can clean this up.

**Comment by charlescng** [comment]
No longer required.
